### ISSUE TITLE
Extended NUnit Project Format

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit-console#147]
The existing nunit project format has not changed since V2. That's handy, since it allows us to open project files that were created with V2, but it doesn't allow us to take advantage of new features.

There's a particular problem with the command-line, which was treated as a sort of virtual project in V2. That is, all the assemblies on the command-line were made into a project internally and could be saved as a project. This was possible in part because of the limitations of the V2 command-line: it was allowed to contain a single project by itself or one or more assemblies. With NUnit 3, it's possible to combile multiple projects and assemblies on the command-line, so any internally created project would have to have the ability to nest projects inside other projects.

Our package format actually does this. The main differences between projects and packages are:
1. Projects support configurations like Debug and Release, while packages do not.
2. Packages support nested subpackages, while projects do not support nested projects.
3. Packages are not currently persisted.

To move forward, I'm proposing we create a new, extended project format, as follows:

1. It would use the file extension `.nunit3`

2. Configurations would be supported but would be optional. That is, assemblies could be directly contained in the project, without need for a Default configuration.

3. Projects could appear anywhere assemblies now appear. They could be of type `.nunit` or `.nunit3.` In the case of `.nunit3` projects, further nesting would be possible.

4. Additional options would be supported as needed. For example, we would allow specifying the config file for an assembly.

5. A new engine extension would have to be created to load projects in this format.

Right now, I'd just like to discuss this idea. If we decide to move ahead, I'll design a format that includes all the options allowed.

BTW, the main reason this came to my mind right now is because I'm trying to figure out how to save the state of the GUI at any given time. Since the gui could have a mix of assemblies and projects loaded all at once, a more flexible format is needed to save the information.

Comments?

**Comment by ChrisMaddock** [comment]
Hi Charlie - can you clarify the need for a `.nunit3` format, rather than just extending the `.nunit` format? Adding an enhancement to allow nested projects sounds like it doesn't need to be a breaking change, unless I'm missing a subtlety?
Would just be nice to limit the number of input formats we support as much as possible!

**Comment by CharliePoole** [comment]
@ChrisMaddock Well... I've proposed a few changes to existing formats lately and our group has generally been somewhat negative about changes considered "breaking." So, in part, this is me adjusting to the new climate, by proposing a new format.

That said, there are a number of things wrong with the existing nunit format - a format I made up myself, without much experience in XML design. One big problem is the fact that settings appear in two places: in a global "Settings" element and as attributes on a Config element. Settings would work much more generally if they were simply attributes of the element that they affect: i.e. either a project or a configuration.

Another problem - one that I think you pointed out yourself - is that the names of elements and attributes are not in a consistent styIe. We could fix that in a new format.

Finally, the new format could be designed in such a way that we could automatically serialize and deserialize it. The design of the current format has always required us to write C# code to deserialize it as an NUnitProject.

An alternative to a new nunit3 format would be a package format. Essentially, we have two internal objects that can represent a group of assemblies: NUnitProject and TestPackage. As I mentioned in the original post, NUnit V2 used to represent the command-line as an NUnitProject. We now represent the command-line as a TestPackage but we don't have a clean way to represent it.

If we decided to make this a TestPackage format, we would merely have to remove the notion of configurations, since a TestPackage represents what you are now running. I thought of calling the file a `.package` file but that struck me as too generic.

Mainly, after all that, I guess it comes down to what I wanted to spend my time on. :smile: I feel that making a backward compatible, enhanced `.nunit` project would require a great deal of work, all of it somewhat adhoc like the present serialization / deserialization code. A new, clean format is a lot easier and would seem to eliminate issues of backward compatibility.
